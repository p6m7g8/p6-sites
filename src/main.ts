import type { Construct } from 'constructs'
import type { IP6CDKWebsiteProps } from 'p6-cdk-website-plus'
import * as fs from 'node:fs'
import * as process from 'node:process'
import * as cdk from 'aws-cdk-lib'
import * as yaml from 'js-yaml'
import { P6CDKWebsitePlus } from 'p6-cdk-website-plus'

// Configuration file
const CONFIG_FILE = 'conf/sites.yml'

/**
 *
 * @param filePath
 * @returns yamlData
 */
function parseYamlFile(filePath: string): IP6CDKWebsiteProps[] {
  const fileContents = fs.readFileSync(filePath, 'utf8')
  const yamlData = yaml.load(fileContents) as IP6CDKWebsiteProps[]
  return yamlData
}

/**
 * Interface for the stack
 * @interface ISite
 * @extends {cdk.StackProps}
 * @property {IP6CDKWebsiteProps} site - The site configuration
 * @property {string} site.hostedZoneName - The hosted zone name
 * @property {string} site.verifyEmail - The email to verify
 * @property {string} site.cloudfrontRecordName - The cloudfront record name
 * @property {cdk.StackProps} env - The environment
 */
interface ISite extends cdk.StackProps {
  site: IP6CDKWebsiteProps
}
/**
 * MyStack
 *
 */
export class MyStack extends cdk.Stack {
  /**
   *
   * @param scope
   * @param id
   * @param props
   */
  constructor(scope: Construct, id: string, props: ISite) {
    super(scope, id, props)

    new P6CDKWebsitePlus(this, props.site.hostedZoneName, props.site)
  }
}

// the AwsEnvironment
const theEnv = {
  account: process.env.CDK_DEPLOY_ACCOUNT || process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEPLOY_REGION || process.env.CDK_DEFAULT_REGION,
}

// create the app and stack
const app = new cdk.App()
const domains: IP6CDKWebsiteProps[] = parseYamlFile(CONFIG_FILE)
domains.forEach((domain) => {
  const stackName = `p6-site-${domain.hostedZoneName.replace(/\./g, '-')}`

  new MyStack(app, stackName, { site: domain, env: theEnv })
})
app.synth()
