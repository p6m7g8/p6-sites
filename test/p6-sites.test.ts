import * as process from 'node:process'
import * as cdk from 'aws-cdk-lib'
import { Template } from 'aws-cdk-lib/assertions'
import { MyStack } from '../src/main'

describe('myStack Snapshot Test', () => {
  it('should match the synthesized CloudFormation template snapshot', () => {
    // the AwsEnvironment
    const theEnv = {
      account: process.env.CDK_DEPLOY_ACCOUNT || process.env.CDK_DEFAULT_ACCOUNT,
      region: process.env.CDK_DEPLOY_REGION || process.env.CDK_DEFAULT_REGION,
    }
    const app = new cdk.App()

    // Create the stack
    const stack = new MyStack(app, 'p6-sites', { env: theEnv })

    // Prepare the stack template for assertions
    const template = Template.fromStack(stack)

    // Assert that the stack matches the snapshot
    expect(template.toJSON()).toMatchSnapshot()
  })
})
