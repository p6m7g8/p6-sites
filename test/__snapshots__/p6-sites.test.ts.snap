// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`myStack Snapshot Test should match the synthesized CloudFormation template snapshot 1`] = `
{
  "Mappings": {
    "AWSCloudFrontPartitionHostedZoneIdMap": {
      "aws": {
        "zoneId": "Z2FDTNDATAQYW2",
      },
      "aws-cn": {
        "zoneId": "Z3RFFRIM2A3IF5",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "p6m7g8orgCachePolicyF5F0FE5F": {
      "Properties": {
        "CachePolicyConfig": {
          "DefaultTTL": 86400,
          "MaxTTL": 31536000,
          "MinTTL": 0,
          "Name": "p6sitesp6m7g8orgCachePolicyC271D8A2-us-east-1",
          "ParametersInCacheKeyAndForwardedToOrigin": {
            "CookiesConfig": {
              "CookieBehavior": "none",
            },
            "EnableAcceptEncodingBrotli": true,
            "EnableAcceptEncodingGzip": true,
            "HeadersConfig": {
              "HeaderBehavior": "none",
            },
            "QueryStringsConfig": {
              "QueryStringBehavior": "none",
            },
          },
        },
      },
      "Type": "AWS::CloudFront::CachePolicy",
    },
    "p6m7g8orgCertificateD5A4BD94": {
      "Properties": {
        "DomainName": "p6m7g8.org",
        "DomainValidationOptions": [
          {
            "DomainName": "p6m7g8.org",
            "ValidationDomain": "p6m7g8.org",
          },
          {
            "DomainName": "www.p6m7g8.org",
            "ValidationDomain": "p6m7g8.org",
          },
        ],
        "SubjectAlternativeNames": [
          "www.p6m7g8.org",
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "p6-sites/p6m7g8.org/Certificate",
          },
        ],
        "ValidationMethod": "EMAIL",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
    "p6m7g8orgCloudfrontDnsRecordAAAAWWW73294712": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "p6m7g8orgDistributionB3F378E3",
              "DomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::FindInMap": [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": "DUMMY",
        "Name": "www.p6m7g8.org.",
        "Type": "AAAA",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "p6m7g8orgCloudfrontDnsRecordWWWF2ADF083": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "p6m7g8orgDistributionB3F378E3",
              "DomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::FindInMap": [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": "DUMMY",
        "Name": "www.p6m7g8.org.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "p6m7g8orgDistributionB3F378E3": {
      "Properties": {
        "DistributionConfig": {
          "Aliases": [
            "www.p6m7g8.org",
          ],
          "Comment": "www.p6m7g8.org",
          "DefaultCacheBehavior": {
            "AllowedMethods": [
              "GET",
              "HEAD",
            ],
            "CachePolicyId": {
              "Ref": "p6m7g8orgCachePolicyF5F0FE5F",
            },
            "Compress": true,
            "TargetOriginId": "p6sitesp6m7g8orgDistributionOrigin16B3C11D3",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Logging": {
            "Bucket": {
              "Fn::GetAtt": [
                "p6m7g8orgLogBucketB30AAD83",
                "RegionalDomainName",
              ],
            },
            "IncludeCookies": true,
          },
          "Origins": [
            {
              "CustomOriginConfig": {
                "OriginProtocolPolicy": "http-only",
                "OriginSSLProtocols": [
                  "TLSv1.2",
                ],
              },
              "DomainName": {
                "Fn::Select": [
                  2,
                  {
                    "Fn::Split": [
                      "/",
                      {
                        "Fn::GetAtt": [
                          "p6m7g8orgMyBucket9D229790",
                          "WebsiteURL",
                        ],
                      },
                    ],
                  },
                ],
              },
              "Id": "p6sitesp6m7g8orgDistributionOrigin16B3C11D3",
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": {
            "AcmCertificateArn": {
              "Ref": "p6m7g8orgCertificateD5A4BD94",
            },
            "MinimumProtocolVersion": "TLSv1.2_2021",
            "SslSupportMethod": "sni-only",
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "p6m7g8orgLogBucketB30AAD83": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AccessControl": "LogDeliveryWrite",
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "OwnershipControls": {
          "Rules": [
            {
              "ObjectOwnership": "ObjectWriter",
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "IgnorePublicAcls": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "p6m7g8orgLogBucketPolicy54549AE9": {
      "Properties": {
        "Bucket": {
          "Ref": "p6m7g8orgLogBucketB30AAD83",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:PutObject",
              "Condition": {
                "ArnLike": {
                  "aws:SourceArn": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:cloudfront::",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":distribution/*",
                      ],
                    ],
                  },
                },
                "StringEquals": {
                  "aws:SourceAccount": {
                    "Ref": "AWS::AccountId",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "delivery.logs.amazonaws.com",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "p6m7g8orgLogBucketB30AAD83",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "p6m7g8orgMyBucket9D229790": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AccessControl": "BucketOwnerFullControl",
        "CorsConfiguration": {
          "CorsRules": [
            {
              "AllowedHeaders": [
                "*",
              ],
              "AllowedMethods": [
                "GET",
                "PUT",
              ],
              "AllowedOrigins": [
                "*",
              ],
              "MaxAge": 3000,
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "IgnorePublicAcls": true,
        },
        "WebsiteConfiguration": {
          "IndexDocument": "index.html",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "p6m7g8orgMyBucketPolicyEFE25436": {
      "Properties": {
        "Bucket": {
          "Ref": "p6m7g8orgMyBucket9D229790",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "p6m7g8orgMyBucket9D229790",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "p6m7g8orgRedirectRedirectAliasRecord0c386926F4066A": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "p6m7g8orgRedirectRedirectDistributionCFDistribution65F58384",
              "DomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::FindInMap": [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": "DUMMY",
        "Name": "p6m7g8.org.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "p6m7g8orgRedirectRedirectAliasRecordSix0c386996DB2583": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "p6m7g8orgRedirectRedirectDistributionCFDistribution65F58384",
              "DomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::FindInMap": [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": "DUMMY",
        "Name": "p6m7g8.org.",
        "Type": "AAAA",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "p6m7g8orgRedirectRedirectBucket87455507": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "WebsiteConfiguration": {
          "RedirectAllRequestsTo": {
            "HostName": "www.p6m7g8.org",
            "Protocol": "https",
          },
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "p6m7g8orgRedirectRedirectCertificateCertificateRequestorFunctionEDDBD27E": {
      "DependsOn": [
        "p6m7g8orgRedirectRedirectCertificateCertificateRequestorFunctionServiceRoleDefaultPolicy83D60229",
        "p6m7g8orgRedirectRedirectCertificateCertificateRequestorFunctionServiceRole35B11BD9",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-346733622331-us-east-1",
          "S3Key": "b073cebcf4d61fb152a30f5a5e57a94df7f980a549fdf1a79a0b18c5750522d8.zip",
        },
        "Handler": "index.certificateRequestHandler",
        "Role": {
          "Fn::GetAtt": [
            "p6m7g8orgRedirectRedirectCertificateCertificateRequestorFunctionServiceRole35B11BD9",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "p6m7g8orgRedirectRedirectCertificateCertificateRequestorFunctionServiceRole35B11BD9": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "p6m7g8orgRedirectRedirectCertificateCertificateRequestorFunctionServiceRoleDefaultPolicy83D60229": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "acm:RequestCertificate",
                "acm:DescribeCertificate",
                "acm:DeleteCertificate",
                "acm:AddTagsToCertificate",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "route53:GetChange",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "route53:changeResourceRecordSets",
              "Condition": {
                "ForAllValues:StringEquals": {
                  "route53:ChangeResourceRecordSetsActions": [
                    "UPSERT",
                  ],
                  "route53:ChangeResourceRecordSetsRecordTypes": [
                    "CNAME",
                  ],
                },
                "ForAllValues:StringLike": {
                  "route53:ChangeResourceRecordSetsNormalizedRecordNames": [
                    "*.p6m7g8.org",
                    "*.p6m7g8.org",
                  ],
                },
              },
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":route53:::hostedzone/DUMMY",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "p6m7g8orgRedirectRedirectCertificateCertificateRequestorFunctionServiceRoleDefaultPolicy83D60229",
        "Roles": [
          {
            "Ref": "p6m7g8orgRedirectRedirectCertificateCertificateRequestorFunctionServiceRole35B11BD9",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "p6m7g8orgRedirectRedirectCertificateCertificateRequestorResource47A43B2A": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DomainName": "p6m7g8.org",
        "HostedZoneId": "DUMMY",
        "Region": "us-east-1",
        "ServiceToken": {
          "Fn::GetAtt": [
            "p6m7g8orgRedirectRedirectCertificateCertificateRequestorFunctionEDDBD27E",
            "Arn",
          ],
        },
        "SubjectAlternativeNames": [
          "p6m7g8.org",
        ],
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "p6m7g8orgRedirectRedirectDistributionCFDistribution65F58384": {
      "Properties": {
        "DistributionConfig": {
          "Aliases": [
            "p6m7g8.org",
          ],
          "Comment": "Redirect to www.p6m7g8.org from p6m7g8.org",
          "DefaultCacheBehavior": {
            "AllowedMethods": [
              "GET",
              "HEAD",
            ],
            "CachedMethods": [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "ForwardedValues": {
              "Cookies": {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": [
            {
              "ConnectionAttempts": 3,
              "ConnectionTimeout": 10,
              "CustomOriginConfig": {
                "HTTPPort": 80,
                "HTTPSPort": 443,
                "OriginKeepaliveTimeout": 5,
                "OriginProtocolPolicy": "http-only",
                "OriginReadTimeout": 30,
                "OriginSSLProtocols": [
                  "TLSv1.2",
                ],
              },
              "DomainName": {
                "Fn::Select": [
                  2,
                  {
                    "Fn::Split": [
                      "/",
                      {
                        "Fn::GetAtt": [
                          "p6m7g8orgRedirectRedirectBucket87455507",
                          "WebsiteURL",
                        ],
                      },
                    ],
                  },
                ],
              },
              "Id": "origin1",
            },
          ],
          "PriceClass": "PriceClass_All",
          "ViewerCertificate": {
            "AcmCertificateArn": {
              "Fn::GetAtt": [
                "p6m7g8orgRedirectRedirectCertificateCertificateRequestorResource47A43B2A",
                "Arn",
              ],
            },
            "SslSupportMethod": "sni-only",
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
